code/parachain/frame/cosmwasm/src/lib.rs

```diff
@@ -498,1084 +513,836 @@ pub mod pallet {
 			Self::refund_gas(outcome, initial_gas, shared.gas.remaining())
 		}
 	}
+}
 
-	impl<T: Config> Pallet<T> {
-		/// Deterministic contract address computation, similar to https://eips.ethereum.org/EIPS/eip-1014.
-		pub(crate) fn derive_contract_address(
-			instantiator: &AccountIdOf<T>,
-			salt: &[u8],
-			code_hash: CodeHashOf<T>,
-			message: &[u8],
-		) -> AccountIdOf<T> {
-			let data: Vec<_> = instantiator
-				.as_ref()
-				.iter()
-				.chain(salt)
-				.chain(code_hash.as_ref())
-				.chain(T::Hashing::hash(message).as_ref())
-				.cloned()
-				.collect();
-			UncheckedFrom::unchecked_from(T::Hashing::hash(&data))
-		}
-
-		/// Deterministic contract trie id generation.
-		pub(crate) fn derive_contract_trie_id(
-			contract: &AccountIdOf<T>,
-			nonce: u64,
-		) -> ContractTrieIdOf<T> {
-			let data: Vec<_> =
-				contract.as_ref().iter().chain(&nonce.to_le_bytes()).cloned().collect();
-			T::Hashing::hash(&data).as_ref().to_vec().try_into().expect(
-				"hashing len implementation must always be <= defined max contract trie id size; QED;",
-			)
-		}
+/// Query cosmwasm contracts
+///
+/// * `contract` the address of contract to query.
+/// * `gas` the maximum gas to use, the remaining is refunded at the end of the transaction.
+/// * `query_request` the binary query, which should be deserializable to `QueryRequest`.
+pub fn query<T: Config>(
+	contract: AccountIdOf<T>,
+	gas: u64,
+	query_request: Vec<u8>,
+) -> Result<QueryResponse, CosmwasmVMError<T>> {
+	let mut shared = Pallet::<T>::do_create_vm_shared(gas, InitialStorageMutability::ReadOnly);
+	let query_request = serde_json::from_slice(&query_request)
+		.map_err(|e| CosmwasmVMError::Rpc(format!("{}", e)))?;
+	Pallet::<T>::sub_level_dispatch(
+		&mut shared,
+		contract.clone(),
+		contract,
+		Default::default(),
+		|vm| {
+			cosmwasm_system_query(vm, query_request)?
```

Comment:
I think this is invalid and we must also hook it right?

Reply:
yes, because this currently doesn't support native contracts.

Reply:
Yes but this is unaltered by this PR